### DISCUSSION TITLE
How to upload files from the server-side?

### DISCUSSION THREAD

**Discussion by rainwashed** [pocketbase/pocketbase#2723] [Q&A] [answered]
I am using the JS pocketbase api on a Node.js server (using Next.js as a framework and their /api folder). Based on personal research I've done, file entries take the use of a ``File`` object, but how would such an object be constructured, especially with the ``File.content``. 

```js
new File(["imgbase64data"] as BlobPart[], "test.jpg", { type: "image/jpg" }
```

but this produces a 

```
ClientResponseError 400: Failed to load the submitted data due to invalid formatting.
    at new ClientResponseError (file:///censored/node_modules/pocketbase/dist/pocketbase.es.mjs:1:2306)
    at file:///censored/node_modules/pocketbase/dist/pocketbase.es.mjs:1:45460
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async create_short (webpack-internal:///(api)/./src/utils/docs/shorts/create_short.ts:18:22)
    at async handler (webpack-internal:///(api)/./src/pages/api/shorts/create.ts:32:21) {
  url: 'http://127.0.0.1:8090/api/collections/shorts/records',
  status: 400,
  response: {
    code: 400,
    message: 'Failed to load the submitted data due to invalid formatting.',
    data: { attachments: [Object] }
  },
  isAbort: false,
  originalError: null
}
```

so what would be the correct formatting for this? The docs do not provide a clear example for this case since it uses the .files of a FormInput.

**Comment by ganigeorgiev** [comment]
Could you provide a more complete code sample showing from where the file data comes from?

Also note that the return response contains a field validation error (eg. mimetype or size check). You can inspect it by wrapping your call with `try/catch` and inspect the `err.response.data` object (_from the above log it seems that the `attachments` field is invalid_).

In general, when uploading files with `multipart/form-data` the field value mist be [`Blob`](https://developer.mozilla.org/en-US/docs/Web/API/Blob) or [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File) instance. Here is an example using node18 and a local file:

```js
formData.append("yourFileField", new Blob([fs.readFileSync("./test.png")]), "test.png");
```

**Reply by rainwashed** [reply]
This file data (currently) is just a fixed Blob object with the Base64 representation of a PNG. I plan to have my own server logic to handle files and convert the uploaded files into the blob object like previously mentioned. 

(this is within a function with the arguments ``text: string``, ``attachments: any[]``, and ``author: string`` (author is taken from another ``Record`` within the users collection, to create references.)
```js
let newDoc: Record = await pb.collection("shorts").create({
            content: sanitize(text),
            attachments: [new Blob('yadayadayada')],
            author: doc.id,
        });
```

**Reply by rainwashed** [reply]
even trying your read file method does not produce anything. my code is
```js
let newDoc: Record = await pb.collection("shorts").create({
            content: sanitize(text),
            attachments: new Blob([
                readFileSync("/home/censored/c/z/src/testing/test_img.jpg"),
            ]),

            author: doc.id,
        });
        ```

**Reply by ganigeorgiev** [reply, marked as answer]
> fixed Blob object with the Base64 representation of a PNG

Make sure that the base64 data is converted properly to its binary representation.

Files upload is currently supported only via `multipart/form-data` requests, meaning that you have to pass a [`FormData` object](https://developer.mozilla.org/en-US/docs/Web/API/FormData). `FormData` is available natively in both browsers and Node.js 18+.

Your code sample can be converted to:

```js
const formData = new FormData();
formData.append("content", sanitize(text))
formData.append("author", doc.id)
formData.append("attachments", new Blob([...]), "test1.txt");
formData.append("attachments", new Blob([...]), "test2.txt"); // repeat if the field supports multiple file uploads

let newDoc = await pb.collection("shorts").create(formData);
```

**Reply by rainwashed** [reply]
that worked, but do you know when it will be updated to just use the object notation? a similar effect for that could be 
``new File([new Blob([...])], "filename")```

**Reply by RRoussev88** [reply]
@ganigeorgiev Is the update API also working with FormData like that?
I have a Users table with `avatar` column that is of type single file and I want my users to be able to change theis avatars.

**Reply by ganigeorgiev** [reply]
@RRoussev88 Yes, both `create` and `update` calls accept the same body data formats.
The plain object notation is also supported with JS SDK v0.17+ (_for an example you can check https://github.com/pocketbase/js-sdk#file-upload_).
